### PR TITLE
Fix SUNDIALS `Context` construction for MPICH

### LINKED ISSUE #2677: Invalid conversion from 'MPI_Comm'
Hi all.

This error appears by compiling `bout++-5.0.0` in Fedora 38+, unless i use `-fpermissive` flag:

```
In file included from /builddir/build/BUILD/bout++-5.0.0/mpich/include/bout/boutcomm.hxx:33,
                 from /builddir/build/BUILD/bout++-5.0.0/mpich/include/bout/field.hxx:42,
                 from /builddir/build/BUILD/bout++-5.0.0/mpich/include/bout/field2d.hxx:34,
                 from /builddir/build/BUILD/bout++-5.0.0/mpich/include/bout/options.hxx:43:
/builddir/build/BUILD/bout++-5.0.0/mpich/src/solver/impls/arkode/arkode.cxx: In constructor 'ArkodeSolver::ArkodeSolver(Options*)':
/usr/include/mpich-x86_64/mpi.h:279:25: error: invalid conversion from 'MPI_Comm' {aka 'int'} to 'void*' [-fpermissive]
  279 | #define MPI_COMM_WORLD ((MPI_Comm)0x44000000)
      |                        ~^~~~~~~~~~~~~~~~~~~~~
      |                         |
      |                         MPI_Comm {aka int}
```
Used:
gcc-13.0.1
mpich-4.0.3
openmpi-4.1.5

This is the full log: https://download.copr.fedorainfracloud.org/results/sagitter/ForTesting/fedora-rawhide-x86_64/05643749-bout++/builder-live.log.gz

### PR DESCRIPTION
**Pull request by dschwoerer**
Resolves #2677

### CODE REVIEW

File: src/solver/impls/arkode/arkode.cxx [hunk 1]
```diff
@@ -226,7 +226,7 @@ ArkodeSolver::ArkodeSolver(Options* opts)
                        .withDefault(false)),
       optimize(
           (*options)["optimize"].doc("Use ARKode optimal parameters").withDefault(false)),
-      suncontext(MPI_COMM_WORLD) {
+      suncontext(static_cast<void*>(MPI_COMM_WORLD)) {
```
Comment (ZedThree) [inline src/solver/impls/arkode/arkode.cxx:229]: This isn't quite the right fix, I'm afraid. For MPICH, `MPI_COMM_WORLD` is an `int` literal, so this is casting that integer value to a random address. We actually need to pass a pointer to the communicator, so `static_cast<void*>&BoutComm::get()`

File: include/bout/sundials_backports.hxx [hunk 2]
```diff
@@ -39,7 +39,7 @@ inline void SUNNonlinSolFree(MAYBE_UNUSED(SUNNonlinearSolver solver)) {}
 #if SUNDIALS_VERSION_MAJOR < 6
 namespace sundials {
 struct Context {
-  Context(MPI_Comm comm MAYBE_UNUSED() = MPI_COMM_NULL) {}
+  Context(void* comm MAYBE_UNUSED() = MPI_COMM_NULL) {}
```
Comment (dschwoerer) [inline include/bout/sundials_backports.hxx:42]: Do we need a default argument? We probably want to make sure we always pass a valid MPI_Comm?
```suggestion
  Context(void* comm MAYBE_UNUSED()) {}
```

Comment (ZedThree) [reply include/bout/sundials_backports.hxx:42]: Yes, the default argument can go

**Review by ZedThree — APPROVED** [review]

**Comment by ZedThree** [comment]
Did this change run on both MPICH and OpenMPI? If so, it can be merged right now, if not, please can someone test it? I guess this is one case where we definitely do want to test across multiple MPI implementations!

**Comment by dschwoerer** [comment]
Yes, I did build it against both MPIs with sundials 6.5.0-1.fc39:

[builder-live.log.gz](https://github.com/boutproject/BOUT-dev/files/11059615/builder-live.log.gz)

With these patches on top of v5.0.0:
Patch:  https://github.com/boutproject/BOUT-dev/pull/2659.patch
Patch:  https://github.com/boutproject/BOUT-dev/pull/2661.patch
Patch:  https://github.com/boutproject/BOUT-dev/pull/2664.patch
Patch:  https://github.com/boutproject/BOUT-dev/pull/2678.patch
Patch:  https://github.com/boutproject/BOUT-dev/pull/2679.patch
Patch:  https://github.com/boutproject/BOUT-dev/commit/489270184ba9e03cb9996c614050cb92513424e6.patch#./petsc_318_compat.patch

**Pull request merged**